tornado/tcpclient.py

```diff
@@ -173,19 +174,30 @@ def connect(self, host, port, af=socket.AF_UNSPEC, ssl_options=None,
         .. versionchanged:: 4.5
            Added the ``source_ip`` and ``source_port`` arguments.
         """
-        addrinfo = yield self.resolver.resolve(host, port, af)
+        if timeout:
```

Comment:
If `timeout` is a `timedelta`, it should be resolved to an absolute time as soon as possible, so it is not used three separate times (allowing  `timeout=timedelta(seconds=5)` to take almost 15 seconds if resolve, connect, and start_tls are all slow).